pkg/kv/store_message.go

```diff
@@ -51,62 +51,133 @@ func (s *StoreMessage) DeleteMsg(ctx context.Context, path string) error {
 	return s.Delete(ctx, []byte(path))
 }
 
-// Scan returns a prefix iterator which returns entries in deserialized format. Scan in store message implementation is msg specific.
-//   Therefore, the prefix given should limit the range of the data to those that can be deserialized by the proto message type.
-func (s *StoreMessage) Scan(ctx context.Context, msgType protoreflect.MessageType, prefix string) (*MessageIterator, error) {
-	return NewMessageIterator(ctx, s.Store, msgType, prefix)
-}
-
-type MessageIterator struct {
-	itr     EntriesIterator
-	msgType protoreflect.MessageType
-	err     error
+func (s *StoreMessage) Scan(ctx context.Context, prefix string) (*PrimaryIterator, error) {
+	return NewPrimaryIterator(ctx, s.Store, prefix)
 }
 
 type MessageEntry struct {
 	Key   string
 	Value protoreflect.ProtoMessage
 }
 
-func NewMessageIterator(ctx context.Context, store Store, msgType protoreflect.MessageType, prefix string) (*MessageIterator, error) {
+func NewMessageEntry(msg protoreflect.MessageType) *MessageEntry {
+	return &MessageEntry{
+		Key:   "",
+		Value: msg.New().Interface(),
+	}
+}
+
+type MessageIterator interface {
+	Next() bool
+	Entry(key *string, value *protoreflect.ProtoMessage) error
+	Err() error
+	Close()
+}
+
+// PrimaryIterator MessageIterator implementation for primary key
+// The iterator iterates over the given prefix and returns the proto message and key
+type PrimaryIterator struct {
+	itr EntriesIterator
+	err error
+}
+
+func NewPrimaryIterator(ctx context.Context, store Store, prefix string) (*PrimaryIterator, error) {
 	itr, err := ScanPrefix(ctx, store, []byte(prefix))
 	if err != nil {
 		return nil, fmt.Errorf("create prefix iterator: %w", err)
 	}
-	return &MessageIterator{itr: itr, msgType: msgType}, nil
+	return &PrimaryIterator{itr: itr}, nil
 }
 
-func (m *MessageIterator) Next() bool {
+func (m *PrimaryIterator) Next() bool {
 	if m.err != nil {
 		return false
 	}
 	return m.itr.Next()
 }
 
-func (m *MessageIterator) Entry() *MessageEntry {
+func (m *PrimaryIterator) Entry(key *string, value *protoreflect.ProtoMessage) error {
```

Comment:
Now I get it. I think we should be aligned with the original store implementation:
```suggestion
func (m *PrimaryIterator) Entry(value *protoreflect.ProtoMessage) (*MessageEntry, error) {
```

Reply:
What we wanted to do here is to avoid unnecessary object creations and pass the responsibility to the user. This way, the user provides us with the pointer to the object and we de-serialize the data to it.
If we already return the data on the func argument - what do you propose we return in the *MessageEntry?